### PR TITLE
Winter 2023 update

### PR DESCRIPTION
**Pull request by eb-64-64**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,54 +1,202 @@
...
+* Pointers aren't magic. They're just regular old C++ variables. In fact,
+  they're just a number that's interpreted as an address in memory.
+* The `*`, `&`, and `->` symbols in C++ aren't magic; all three symbols
+  represent a specific operation to do with pointers, and `*` and `&` both
+  additionally enable you to name certain types in C++.
+* Both `*` and `&` can be used in C++ to name types. `T*` names a
+  pointer to a `T`, while `T&` names a reference to a `T`.
+* Both `*` and `&` can additionally be used as an operator. `*v` dereferences
+  the pointer `v`, while `&v` takes the address of `v`.
+* `v->data_member` is just a shorthand for `(*v).data_member`.
```
Comment (byubean) [inline README.md:288]: Most students should know this already, so I'm not sure we need it in the main document. Can you please move this to a separate `.md` file (maybe `tips-from-the-tas.md`)? Then add a link to it in the Readme; something like
> Additional tips on C++ and the BST can be found in [Tips From the TAs](tips-from-the-tas.md)

File: README.md [hunk 2]
```diff
@@ -1,54 +1,202 @@
...
+  method may result in a different tree. In order to make sure that your tree
+  matches the tree the tests expect:
+  * When removing a node with only one child, replace that node with
+  its child. 
+  * When removing nodes with two children, use the in-order predecessor
+  of the node to be removed.
+
+### Part 4 - Clear (15 points)
+* Implement the `clear()` method from `BSTInterface`.
+
```
Comment (byubean) [inline README.md:68]: Can you add a section called "Testing your code" that explains the test files and how to interpret the test output (both the pretty and ugly trees). Also explain the other resources we provide in the repo (e.g. scratch.cpp)

File: key_file2.txt [hunk 3]
```diff
@@ -1,70 +1,26 @@
...
+ 1   3   5   7   9  11  13  15
+
+tree.contains(14) = true
+tree.contains(4) = true
+tree.contains(5) = true
+tree.contains(15) = true
+tree.contains(2) = true
+tree.contains(-3) = false
+tree.contains(17) = false
+tree.contains(35) = false
```
Comment (byubean) [inline key_file2.txt]: Let's leave out 7 and test for it (`tree.contains(7)`)

File: key_file3.txt [hunk 4]
```diff
@@ -1,197 +1,189 @@
...
+               8
+
+       4              12
+
+   2       6      10      14
+
+ 1   3   5   7   9  11  13  15
+
+Attempting to remove 6 from the tree...true
+Pretty printing the tree...
```
Comment (byubean) [inline key_file3.txt]: Remove leaf: ✅

File: key_file3.txt [hunk 5]
```diff
@@ -1,197 +1,189 @@
...
+               8
+
+       4              12
+
+   2       5      10      14
+
+ 1   3       7   9  11  13  15
+
+Attempting to remove 5 from the tree...true
+Pretty printing the tree...
```
Comment (byubean) [inline key_file3.txt]: Remove node with only right child: ✅

File: key_file3.txt [hunk 6]
```diff
@@ -1,197 +1,189 @@
...
+Attempting to remove 7 from the tree...true
+Pretty printing the tree...
+
+               8
+
+       4              12
+
+   2              10      14
+
+ 1   3           9  11  13  15
```
Comment (byubean) [inline key_file3.txt]: Remove leaf: not necessary. You can remove this test case.

File: key_file3.txt [hunk 7]
```diff
@@ -1,197 +1,189 @@
...
+
+Attempting to remove 2 from the tree...true
+Pretty printing the tree...
```
Comment (byubean) [inline key_file3.txt:53]: Remove node with two leaf children: ✅

File: key_file3.txt [hunk 8]
```diff
@@ -1,197 +1,189 @@
...
+               8
+
+       4              12
+
+   1              10      14
+
+     3           9  11  13  15
+
+Attempting to remove 1 from the tree...true
+Pretty printing the tree...
```
Comment (byubean) [inline key_file3.txt]: Another leaf: we can skip this one.

File: key_file3.txt [hunk 9]
```diff
@@ -1,197 +1,189 @@
...
+               8
+
+       4              12
+
+   3              10      14
+
+                 9  11  13  15
+
+Attempting to remove 9 from the tree...true
+Pretty printing the tree...
```
Comment (byubean) [inline key_file3.txt]: Skip

Comment (byubean) [reply key_file3.txt]: Remove 12 at this step. This will test their logic for finding the in-order predecessor.

Comment (byubean) [reply key_file3.txt]: Then test for removing 1, 7, 12, and 18 (which are all absent).

Comment (byubean) [reply key_file3.txt]: Finally, do a quick removal of all the nodes (no pretty printing, just true/false; just repeatedly remove the root) to see that they handle the empty tree scenario correctly.

File: scratch.cpp [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+// easier to read if the tree is not too tall and doesn't have too big of
+// numbers.
+void ugly_print_tree(const BST &tree) {
+    Node *root = tree.getRootNode();
+    if (root == nullptr) {
+        std::cout << "Empty tree" << std::endl;
+    } else {
+        ugly_print_tree_helper(root);
+    }
+}
```
Comment (byubean) [inline scratch.cpp]: Consider moving this code (here and in `test.cpp` to a header file that can be included in both places.

**Review by byubean — COMMENTED** [review]
We had talked about having tests that would work well for AVL, but that's too complicated right now. These tests look pretty good (just a few comments).

Good work!

File: README.md (hunk 1 above)
Comment (eb-64-64) [reply README.md:288]: Yeah, this felt a bit out of place. I've just seen quite a few students putting random asterisks in their code, trying to make it compile, so this felt needed.

Comment (eb-64-64) [reply README.md:288]: Do you want the whole thing in another file, or just the part on pointers (leaving the passing pointers by reference part in the main document)?

**Comment by byubean** [comment]
Everything from comment downward

File: README.md (hunk 1 above)
Comment (eb-64-64) [reply README.md:288]: I'm sorry, I'm still a little confused. This bit that I wrote has two logical sections: one, recommending passing pointers by reference; and two, explaining C++ pointer syntax. The highlighted section in your review starts on line 103, which is halfway through the explanation of passing pointers by reference. In fact, the highlighted piece cuts a sentence in half. Would you like both sections moved, or only the section on C++ pointer syntax?

File: key_file3.txt (hunk 9 above)
Comment (eb-64-64) [reply key_file3.txt]: I'm a bit confused by your feedback on the tests, as 6 (the first remove) isn't a leaf and if we skip removing 1 and 7, they won't be absent as you stated above. I wonder if reworking the tests would be best. I propose the following:

Start with the original tree that looks like this:
```
               8

       4              12

   2       6      10      14

 1   3   5   7   9  11  13  15
```

Then remove the following nodes:
5 -- leaf node
6 -- node with only a right child
3 -- another leaf node, but it'll set up for the next remove
2 -- node with only a left child
4 -- node with two leaf children
12 -- testing the logic of finding an inorder predecessor

After having hit those cases, test some absent values:
3
12
0
18

Finally, repeatedly remove the head until empty:
8
7
1
11
10
9
14
13
15

**Review by byubean — APPROVED** [review]

**Pull request merged**